Header extension JEP (#1)
imonteroperez/jep: jep/401/README.adoc

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Draft :speech_balloon:

| Type
| Standards

| Created
| 2021-11-28

//
//
// Uncomment if there is an associated placeholder JIRA issue.
//| JIRA
//| :bulb: link:https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
//
//
// Uncomment if there will be a BDFL delegate for this JEP.
//| BDFL-Delegate
//| :bulb: Link to github user page :bulb:
//
//
// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
//| Discussions-To
//| :bulb: Link to where discussion and final status announcement will occur :bulb:
//
//
// Uncomment if this JEP depends on one or more other JEPs.
//| Requires
//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
//
//
// Uncomment and fill if this JEP is rendered obsolete by a later JEP
//| Superseded-By
//| :bulb: JEP-NUMBER :bulb:
//
//
// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

Jenkins does not provide a customization mechanism for header.

> Unique existing approach based on the https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin] has a limited capabilities (see Specification section for more details). It makes it difficult: (i) to provide branding capabilities to include custom logos, styles or other elements, and (ii) to make feasible to rewrite some functionality or include additional one to some elements like the search bar. Sometimes, this limited customization capabilities implies a barrier on Jenkins adoption inside enterprises. Having the ability to customize the header (not only from the UI POV) will help to avoid that situation.

Reviewer: https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line otherwise it is very hard for reviewers to comment on particular phrases or offer suggestions, and diffs will be opaque.

Author: Updated in https://github.com/imonteroperez/jep/pull/1/commits/ccb27a8886ce7150f7b6f8cba7dc8c94c5e63090

This proposal provides an approach to make it extensible to provide a customization mechanism for a better integration that also reduces current tech debt.

== Specification

The main aspect of the change is the introduction of a new extension point `Header` as an abstract class that provides capabilities to render a specific header and a default implementation of that, named `JenkinsHeader` that is enabled by default always.

> All headers will provide an implementation of a priority method, via https://javadoc.jenkins.io/hudson/Extension.html[ordinal] though the `Extension` annotation, that will help to override, based on its value, which header will be rendered. Default one will provide a `Integer.MIN_VALUE` value.

Reviewer 2: > All headers will provider an implementation of a priority method

Do you mean in addition to the `ordinal` definition in the annotation? Your examples below do not include a "priority method".

---

> via ordinal

From the [Extension JavaDoc](https://javadoc.jenkins.io/hudson/Extension.html#ordinal--)

> This is a rather poor approach to the problem, so its use is generally discouraged.

At this point I do not understand why it's discouraged, from my PoV it seems fine here.

Author: Yes, I was talking about `optional`. I will update the content of the doc to make it crystal clear and to avoid missunderstandings

Author: Updated on https://github.com/imonteroperez/jep/pull/1/commits/b59850a8b9414a30b9250114c934b91403309977, removed the word `method` to avoid confusion on implementing a new method.

> On `pageHeader.jelly` file we will perform a refactor to make it more modular providing different sections:

Reviewer 2: > more modular providing different sections

Could you elaborate more on this point?

IIUC you plan to provide the default implemention of those fragments at the `Header` level to let any implemention relies on them at will? If you plan to provide them with JenkinsHeader only, that could break the expectations of reusability because in such case you mush inherits from JenkinsHeader.

I would say this will be clearer with the actual PR.

Reviewer 2: Looking at [core PR's headerContent](https://github.com/jenkinsci/jenkins/commit/59f68357c05f9c3d094ca835efea2212c5e38473#diff-f453e7103580ebb72407cc4f22cd6c75de30ef2362255567905d13080bf82c6c) it seems that you want to put the fragment at the Header level.

> * `headerDocumentation.jelly`: that will provide the content of the documentation of the Jelly page

Reviewer 2: By documentation, do you mean the `<st:documentation>` part? I don't think it will work. The purpose of the st:documentation is to generate JavaDoc associated with the taglib you are creating. Doing it dynamically will add no value compared to a comment with explanations.

💡 See the LayoutTabLib.gdsl for example (after a `mvn:compile`)

Reviewer 2: Looking at the [reference implemention](https://github.com/imonteroperez/custom-header-plugin/blob/main/src/main/resources/org/jenkinsci/plugins/custom/header/CustomHeader/headerDocumentation.jelly), it's meant exactly for this. So my question there, what is your expectations? :) I haven't tested but I am 90% sure it's not working as we can expect.

What you are describing in the documentation is the "taglib signature" of `pageHeader.jelly`, not something related to the header. The implementation of the extension do not have an impact on which variable are passed to the taglib.

💡 So suggestion here: keep the documentation at the top of the pageHeader.jelly.

Author: Good catch! I did not try to generate the JavaDoc on my testing. I will try and confirm your concern and update it properly

Reviewer: It is inappropriate to be discussing detailed technical considerations of a proposed API in an OSS component in a private forum.

Author: Removed `headerDocumentation` content in https://github.com/imonteroperez/jep/pull/1/commits/f3f0da4a9672c23b12500d0bb4aad4bff172b99a. Also tested in: https://github.com/jenkinsci/jenkins/commit/774139b227058ae55ba45178c3abd82002f9abd7

> It is inappropriate to be discussing detailed technical considerations of a proposed API in an OSS component in a private forum.

I will create a PR as reference implementation for this.

* `preHeader.jelly`: that will provide the content of the elements that want to be provided/rendered before header div is rendered on the browser
* `headerContent.jelly`: that will provide the content of the header div
* `postHeader.jelly`: that will provide the content of the elements that want to be provided/rendered after the header div is rendered on the browser

Thus, `pageHeader.jelly` should look like:

> ```

Reviewer: ```suggestion
```xml
```

<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
	<j:set var="header" value="${h.header()}" />

> 	<st:include it="${header}" page="headerDocumentation.jelly"/>

Reviewer: ```suggestion
	<st:include it="${header}" page="headerDocumentation"/>
```
to leave the choice of implementation language up to the implementor. But these sorts of details are probably best left for an actual code PR, which I would advise filing (in draft form) sooner rather than later

	<st:include it="${header}" page="preHeader.jelly"/>
	<st:include it="${header}" page="headerContent.jelly"/>
	<st:include it="${header}" page="postHeader.jelly"/>
</j:jelly>
```

> The `h.header()` method is provided in the `Functions` and it will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value). In case two or more headers are provided via external plugins with the same priority (and max), then the default one will be provided and a warning message would be provided on the logs.

Reviewer 2: 💡 Do we really want to add a new method in a class that has already 2200+ lines of code? :)

Please do not reply we could put that code in Jenkins, as the `app` variable is present in jelly views, it has already 5600+ lines :trollface: 

Proposal: add a static method inside Header and call this from the view directly. (you can look for `j:invokeStatic` for examples). That could have the advantage to keep that code local instead of spreading it to Functions where people could use it while it's not expected to be used outside of `pageHeader.jelly` IIUC.

Author: Great feedback here! I was not happy including things on `Functions`. Tested on https://github.com/jenkinsci/jenkins/commit/63367c7b65447a89ab55136d43fca18e65769a0a and working as expected. I will update the content of the JEP now to reflect this.

Author: JEP content updated in: https://github.com/imonteroperez/jep/pull/1/commits/86598276006f342898245b701ce88c74a88e67b2

> In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header. As a follow up, any given implementation of the extension can provide custom extension points for further or more granular customization capabilities.

Reviewer 2: > [...] simplificy [...]

I think your approach with 4 different fragments is already an over-engineered approach at this point :)

Author: Reduced to three. Just looking for a modular view of current pageHeader.jelly :-)

Author: Reduced to one :-)

=== Extensibility

Jenkins users will be able to interact with the `Header` extension as follows in case want to customize the header of their instances by creating a custom plugin. For that purpose, they need just to extend `Header` and provide an implementation of the priority method using the ordinal value to specify desired priority value.

> == Motivation

Reviewer: I think it would be appropriate here to discuss the fact that CB is currently using https://github.com/stephenc/diffpatch-maven-plugin as a workaround but that this is unsatisfactory. Add a status quo section under **Reasoning** that goes into detail.

Author: Added and updated reasoning section

Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.

There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.

> == Reasoning

Reviewer: Misusing this section I think. It should discuss alternatives that were considered and rejected—why you picked this style and not another. https://github.com/jenkinsci/jep/blob/master/jep/1/README.adoc#required-sections

Author: Missed sharing why I discarded the other approaches. Now updated.

We will use an example of a Jenkins user that wants to update Jenkins header through an ad hoc Jenkins plugin that follows the principles provided on Specification section:

=== Jenkins core

* Let’s consider the following definition of the `Header` on: `core/src/main/java/jenkins/views/Header.java`

```
package jenkins.views;

import hudson.ExtensionPoint;

> public abstract class Header implements ExtensionPoint {
> 
>     /**
>      * Check if the header is enabled. By default it is if installed,
>      * but the logic is deferred in the plugins.
>      * @return
>      */
>     public abstract boolean isHeaderEnabled();

Reviewer: or
```suggestion
public interface Header extends ExtensionPoint {

    /**
     * Check if the header is enabled. By default it is if installed,
     * but the logic is deferred in the plugins.
     * @return
     */
    boolean isHeaderEnabled();
```

}
```

* Let’s consider the following implementation of the Jenkins header on: `core/src/main/java/jenkins/views/JenkinsHeader.java`

```
package jenkins.views;

import hudson.Extension;

@Extension(ordinal = Integer.MIN_VALUE)
public class JenkinsHeader extends Header {

    @Override
    public boolean isHeaderEnabled() {
        return true;
    }
}
```

* As mentioned before, the `Functions` method `header()` will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value)

```
[...]
@Restricted(NoExternalUse.class)
@CheckForNull
public static Header header() {
    List<Header> headers = ExtensionList.lookup(Header.class).stream()
                .filter(header -> header.isHeaderEnabled())
                .collect(Collectors.toList());
    if (headers.size() > 0) {
        if (headers.size() > 1) {
            LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
        } else {
            return headers.get(0);
        }
    }

>     return new JenkinsHeader();

Reviewer 2: AFAICT this code is looking for all extensions, not just the ones from plugins as described above by

> [...] In case two or more headers are provided via external plugins with the same priority (and max), then the default one will be provided and a warning message would be provided on the logs.

As JenkinsHeader seems to be part of the the `headers`, if you provide any other implemention, you will get the warning + the JenkinsHeader.

⁉️ I have not tested it manually, what was the result of your tests?

Suggestion: just return the one with higher priority.

Reviewer 2: OK I read the [reference implementation](https://github.com/imonteroperez/custom-header-plugin/blob/main/src/main/java/org/jenkinsci/plugins/custom/header/CustomHeaderExtensionFilter.java). As you have to use a `CustomHeaderExtensionFilter`, it means that this code is not working as expected ;) when we are proposing an extension, I would say that we should provide something that is working easily for the developers, and thus, not requirement "advance" knowledge like the usage of an ExtensionFilter to ensure it's working as expected.

Author: > Suggestion: just return the one with higher priority

IIRC `ExtensionList.lookup` takes into account the optional value to order it in a transparent manner.

> we should provide something that is working easily for the developers, and thus, not requirement "advance" knowledge like the usage of an ExtensionFilter to ensure it's working as expected.

I see your point but at the same time we want to limit the number of active implementations to one. cc/ @fcojfernandez WDYT?

Author: > As JenkinsHeader seems to be part of the headers, if you provide any other implemention, you will get the warning + the JenkinsHeader. [...] I have not tested it manually, what was the result of your tests?

If you don't provide an ExtensionFilter, yes, it will show the WARNING on the log and render the default Jenkins one.

}
```

* Once we launch Jenkins with the proposed changes on the core, we will obtain the expected/current header working without any issue

=== Custom UI plugin

* Create a new plugin following the usual procedure
* Provide an implementation of the custom Header (i.e: `src/main/java/org/jenkinsci/plugins/custom/header/CustomHeader.java`)

```
[...]

> @Extension(optional = true, ordinal = Integer.MAX_VALUE)

Reviewer 2: 💡 I would suggest to avoid using `MAX_VALUE` as it will be easily copy-pasted by other people without understanding that if they do not provide the associated systemProperty, it will be a PITA for administrators.

Having a value like 100 should be largely good enough.

Reviewer 2: ```suggestion
@Extension(ordinal = Integer.MAX_VALUE)
```

The `optional` usage seems incorrect here. Optional is meant for extension that cannot be loaded due to class loading error, and thus, their thrown exception is muted. In your case (and the reference impl), there is no need for this.

💡 So I would suggest to just remove it to prevent adding confusion on something that is already complicated.

If you thought about a link between `optional` and the `isHeaderEnabled()`, there is none (or I am missing a point)

public class CustomHeader extends Header {

    @Override
    public boolean isHeaderEnabled() {
        // Disable/enable the header based on an ENV var

>         boolean isDisabled = "true".equalsIgnoreCase(System.getenv("CUSTOM_HEADER_DISABLE"));

Reviewer 2: ⁉️ Are you sure you want to use getenv and not a "regular" system property?
env = environment variable of the system where you run Jenkins
systemProperty = -Dxxx=yyy variable you pass as argument when you start the war

From my PoV the `ENV` approach as mentioned in the comment is meant for CloudBees docker image, but not necessarily as an example for regular plugins, and thus I would suggest to just put a system property for the public JEP while using `getenv` in the CB implementation.

(yes I am scared by the people (me included) who just copy-paste stuff :D)

Author: Yep, it was just an example, but I update it to take into account your comments. See: https://github.com/imonteroperez/jep/pull/1/commits/b59850a8b9414a30b9250114c934b91403309977 
```
    @Override
    public boolean isHeaderEnabled() {
        // Disable/enable the header based on an ENV var and/or system property
        boolean isDisabled = System.getProperty(CustomHeader.class.getName() + ".disable") != null ?
                "true".equalsIgnoreCase(System.getProperty(CustomHeader.class.getName() + ".disable")) :
                "true".equalsIgnoreCase(System.getenv("CUSTOM_HEADER_DISABLE"));
        return !isDisabled;
    }
```

        return !isDisabled;
    }
}
```

* Provide the jelly files to override the core ones: headerContent, headerDocumentation, preHeader and/or postHeader. For that purpose, use the common location convention. For the previous example: `src/main/resources/org/jenkinsci/plugins/custom/header/CustomHeader/`.
* See the sample implementation provided in the Reference Implementation section.

== Backwards Compatibility

Existing headers will continue to work as expected

== Security

No specific security considerations

== Infrastructure Requirements

No impact on the Jenkins project infrastructure

== Testing

> There are no testing issues related to this proposal

Reviewer: Well, probably there are.

For example, @ahenryjard (IIRC) and others mentioned that writing tests specific to the header using a patched core are currently difficult because of PCT in `unified-release@next`. So you can discuss how a proposed solution will solve these issues: if a customized SDA header is an extension in a plugin then having this plugin on your test classpath will suffice to let UI tests run in the expected way, regardless of core provenance.

Author: You are right. Updated content.

== Reference Implementation

> * Proposed changes on Jenkins core: https://github.com/jenkinsci/jenkins/commit/59f68357c05f9c3d094ca835efea2212c5e38473

Reviewer: Ought to be https://github.com/jenkinsci/jenkins/compare/jenkins-header-revamp but there is something wrong with how you set up the base branch I think?

Author: Removed branch `jenkins-header-revamp` from `jenkinsci/jenkins`. Using my own fork here: https://github.com/jenkinsci/jenkins/compare/master...imonteroperez:header-revamp?expand=1. Updated reference in: https://github.com/imonteroperez/jep/pull/1/commits/b59850a8b9414a30b9250114c934b91403309977

Also, other alternatives has been evaluated as workaround like the https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] with no satisfactory results. See Reasoning section for futher details.

So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.

== Reasoning

We will use an example of a Jenkins user that wants to update Jenkins header through an ad hoc Jenkins plugin that follows the principles provided on Specification section:

=== Jenkins core

* Let’s consider the following definition of the `Header` on: `core/src/main/java/jenkins/views/Header.java`

```
package jenkins.views;

import hudson.ExtensionPoint;

public interface Header extends ExtensionPoint {

    /**
     * Check if the header is enabled. By default it is if installed,
     * but the logic is deferred in the plugins.
     * @return
     */
    boolean isHeaderEnabled();

}
```

> * Let’s consider the following implementation of the Jenkins header on: `core/src/main/java/jenkins/views/JenkinsHeader.java`

Reviewer: Just file a reference implementation PR and link to it. A JEP is not a good place for code review.

| Customizable Jenkins header

| Sponsor
| link:https://github.com/imonteroperez[Ildefonso Montero]

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Draft :speech_balloon:

| Type
| Standards

| Created
| 2021-11-28

//
//
// Uncomment if there is an associated placeholder JIRA issue.
//| JIRA
//| :bulb: link:https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
//
//
// Uncomment if there will be a BDFL delegate for this JEP.
//| BDFL-Delegate
//| :bulb: Link to github user page :bulb:
//
//
// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
//| Discussions-To
//| :bulb: Link to where discussion and final status announcement will occur :bulb:
//
//
// Uncomment if this JEP depends on one or more other JEPs.
//| Requires
//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
//
//
// Uncomment and fill if this JEP is rendered obsolete by a later JEP
//| Superseded-By
//| :bulb: JEP-NUMBER :bulb:
//
//
// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

> Jenkins does not provide a customization mechanism for header.

Reviewer 3: NIT: I would add some explanations on why this is bad, for example related to jenkins adoption inside enterprises. How having the ability to customize the header is appealing to big corporations or how can be used to effectively diferentiate between production and testing instances, just to name a few use cases

Author: Included on https://github.com/imonteroperez/jep/pull/1/commits/f559f2dd07ecdd27de98ddacc77b78ca972ac197

Unique existing approach based on the https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin] has a limited capabilities (see Specification section for more details). It makes it difficult: (i) to provide branding capabilities to include custom logos, styles or other elements, and (ii) to make feasible to rewrite some functionality or include additional one to some elements like the search bar.

> This proposal provides an approach to make it extensible to provide a customization mechanism for a better integration that also reduces current tech debt.

Reviewer 3: Not sure if this is correct

Author: Do you mind elaborating a little bit on your argument, please? Thanks in advance!

Reviewer: Sentence needs to be reworded for clarity.

Author: Updated in https://github.com/imonteroperez/jep/pull/1/commits/ccb27a8886ce7150f7b6f8cba7dc8c94c5e63090

* `headerContent.jelly`: that will provide the content of the header div
* `postHeader.jelly`: that will provide the content of the elements that want to be provided/rendered after the header div is rendered on the browser

Thus, `pageHeader.jelly` should look like:

```
<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
	<j:set var="header" value="${h.header()}" />
	<st:include it="${header}" page="headerDocumentation.jelly"/>
	<st:include it="${header}" page="preHeader.jelly"/>
	<st:include it="${header}" page="headerContent.jelly"/>
	<st:include it="${header}" page="postHeader.jelly"/>
</j:jelly>
```

The `h.header()` method is provided in the `Functions` and it will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value). In case two or more headers are provided via external plugins with the same priority (and max), then the default one will be provided and a warning message would be provided on the logs.

> In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header. As a follow up, each mentioned section of the header could have its own extension to override only a specific part of the header.

Reviewer 3: ```suggestion
In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header. As a follow up, any given implementation of the extension can provide custom extension points for further or more granular customization capabilities.
```

=== Extensibility

> Jenkins users will be able to interact with the `Header` extension as follows in case want to customize the header of their instances via plugin:

Reviewer 3: ```suggestion
Jenkins users will be able to interact with the `Header` extension as follows in case want to customize the header of their instances by creating a custom plugin:
```

> * Extend `Header` and provide an implementation of the priority method giving to this header the max possible value (via ordinal value)

Reviewer 3: Nit: If I have to follow this spec then all plugins will end using `Integer.MAX_VALUE` making the Ordinal system basically a no op.

I would rewrite to not suggest any value.

```suggestion
* Extend `Header` and provide an implementation of the priority method using the ordinal value to specify desired priority value
```

> * Use an `ExtensionFilter` that would make it impossible to provide any additional Header instance

Reviewer 4: Beware of that: You're telling our customers how to get rid of our own header, which goes against any try to protect our own brand. I wouldn't mention the `ExtensionFilter` and let others to figure out how to do it. 
PS: trying to mention Runxia to open the discussion, but I cannot find her

Reviewer 3: I agree with @fcojfernandez

Author: Got it, I will remove this part

Author: Removed in https://github.com/imonteroperez/jep/pull/1/commits/b240f8f85965918e0fd775ecfad4fe2121fe555b

== Motivation

Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.

> There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, it the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.

Reviewer 4: > then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts

is this too close to internal information? I'd check if it's safe to say that before publishing it.

Reviewer 4: typo (apart from my previous comment)
```suggestion
There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.
```

> In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration

for me, this is the key point

=== Custom UI plugin

* Create a new plugin following the usual procedure
* Provide an implementation of the custom Header (i.e: `custom-ui-plugin/src/main/java/foo/bar/CustomHeader.java`)

```
[...]
@Extension(optional = true, ordinal = Integer.MAX_VALUE)
public class CustomHeader extends Header {

    @Override
    public boolean isHeaderEnabled() {
        // Disable/enable the header based on an ENV var
        boolean isDisabled = "true".equalsIgnoreCase(System.getenv("CUSTOM_HEADER_DISABLE"));
        return !isDisabled;
    }
}
```

* Use an `ExtensionFilter` to avoid other `Header` to be available (i.e.: `custom-ui-plugin/src/main/java/foo/bar/CustomHeaderExtensionFilter.java`)

```
[...]
@Extension
public class CustomHeaderExtensionFilter extends ExtensionFilter {

    @Override
    public <T> boolean allows(Class<T> type, ExtensionComponent<T> component) {
        boolean isCustom = component.getInstance().getClass().equals(CustomHeader.class);
        boolean isForbiddenHeader = type.equals(Header.class) && !isCustom;
        return !isForbiddenHeader;
    }
}

> ```

Reviewer 4: ditto: dangerous to share what we plan for the header

Author: Same than previous. I will remove this.

Author: Removed in https://github.com/imonteroperez/jep/pull/1/commits/b240f8f85965918e0fd775ecfad4fe2121fe555b

> * Update jelly files (headerContent, headerDocumentation, preHeader and/or postHeader)

Reviewer 4: How those jelly files are included?

Author: Updated on https://github.com/imonteroperez/jep/pull/1/commits/f559f2dd07ecdd27de98ddacc77b78ca972ac197 to describe how these files are included. Mainly using the usual location convention.
